pkg/provider/aws/client.go

```diff
@@ -164,20 +174,21 @@ func NewAwsClient(profile, region, configFile string) (Client, error) {
 	}
 
 	// Validate the creds
-	_, err = awsClient.GetCallerIdentity(&sts.GetCallerIdentityInput{})
-	if aerr, ok := err.(awserr.Error); ok {
-		switch aerr.Code() {
-		case "InvalidClientTokenId":
-			if profile != "" {
-				klog.Error(fmt.Sprintf("Profile %s provided has invalid credentials. Please validate them and try again. Note this does not indicate a problem with the cluster's credentials", profile))
-			} else {
-				klog.Error("Credentials provided to osdctl are invalid. Please validate them and try again. Note this does not indicate a problem with the cluster's credentials")
+	if _, err := awsClient.GetCallerIdentity(nil); err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case "InvalidClientTokenId":
+				if region == endpoints.UsGovEast1RegionID || region == endpoints.UsGovWest1RegionID {
+					return nil, fmt.Errorf("failed `aws sts get-caller-identity` validation: %v", err)
+				}
+				klog.Infoln("credentials provided invalid, trying GovCloud by setting region to us-gov-west-1.")
+				return NewAwsClient(profile, endpoints.UsGovWest1RegionID, configFile)
```

Comment:
This might have the potential for infinite recursion if the profile or configFile is invalid, or is a missing region the only reason for the InvalidClientTokenId error?

Reply:
An incorrect region (since it defaults to us-east-1) isn't the only reason; invalid values in the provided AWS credentials profile or configFile are other reasons as you mentioned.

However, if it runs into `InvalidClientTokenId` again after the recursive call, it would get caught by line 181 the next time it comes around and exit out since this recursive call explicitly sets it to us-gov-west-1. If not, it'll error out of the default case in line 187.

Reply:
Ah gotcha, good point. Thanks!